airflow/models/dag.py

```diff
@@ -3177,11 +3179,11 @@ def bulk_write_to_db(
                 stored_dataset.is_orphaned = expression.false()
                 stored_datasets[stored_dataset.uri] = stored_dataset
             else:
-                session.add(dataset)
-                stored_datasets[dataset.uri] = dataset
-
-        session.flush()  # this is required to ensure each dataset has its PK loaded
```

Comment:
Is it possible that this flush is not only responsible for ensuring the dataset PK availability, but for something more and that results into tests flakiness?